torch/csrc/dynamo/eval_frame.c

```diff
@@ -314,31 +396,228 @@ static void destroy_cache_entry(CacheEntry* e) {
   free(e);
 }
 
-inline static CacheEntry* get_cache_entry(PyCodeObject* code) {
-  CacheEntry* extra = NULL;
+typedef struct {
+  PyObject_HEAD
+  CacheEntry* cache_entry;
+} CacheEntryWrapper;
+
+static PyTypeObject CacheEntryWrapperType = {
+  PyVarObject_HEAD_INIT(NULL, 0)
+  .tp_name = "torch._C.dynamo.eval_frame.CacheEntryWrapper",
+  .tp_basicsize = sizeof(CacheEntryWrapper),
+  .tp_itemsize = 0,
+  .tp_flags = Py_TPFLAGS_DEFAULT,
+  .tp_new = PyType_GenericNew,
+};
+
+
+// The cache lives on the extra segment of code object. It can have one of the 4 possible values
+// 1) NULL - First time accessed the extra segment.
+// 2) SKIP_CODE - We will skip the original frame.
+// 3) CacheEntryWrapper - This contains the cache_entry for the frame.
+// 4) A dict from nn module to the CacheEntryWrapper.
+//
+// Extracting a cache entry is split into two functions -
+// get_cache_entry_without_unwrapping and unwrap_cache_entry. The reason for
+// this split is to enable early return for SKIP_CODE.
+
+// For (1) and (2), we need to short circuit and return quickly for skipped
+// frames. Therefore, we cover this separately in get_cache_entry_without_unwrapping.
+
+// For (3) - If the frame is not a method of a nn.Module instance,
+// we just store the CacheEntry object directly on the extra segment.
+// For (4) - If the frame is a method of a nn.Module instance, we
+// store the CacheEntry for each nn module instance. This ensures that if a
+// SubModule is instantiated multiple times in a nn.Module and there is a
+// graph break in the SubModule, we save the CacheEntry per submodule isntance
+// and not per code object. If it is per code object (which is static and
+// shared across all nn module instances), this leads to collisions.
+
+inline static PyObject* get_cache_entry_without_unwrapping(PyCodeObject* code) {
+  // This just returns the extra value at cache entry index. It does not do any
+  // unwrapping mentioned in (3) and (4). This is useful to check if the frame
+  // has to be skipped and we can return quickly in _custom_eval_frame. This is
+  // useful in cases where we do lots of skips (like pytorch tests).
+   PyObject* extra = NULL;
   _PyCode_GetExtra((PyObject*)code, cache_entry_extra_index, (void*)&extra);
   return extra;
 }
```

Comment:
The preferable way to do this is to have your layers of logic match a structure. Instead of entry_without_unwrapping, just return an object with a `skipped` bool that you can pass to an `unwrap` function, which then gives you a cache entry?